[Bug] [IPP Feedback Survey] Error handling in `GetIPPFeedbackBannerData`

samiuelson opened the pull request:
This PR fixes error handling in `GetIPPFeedbackBannerData` use case. 

<!-- Remember about a good descriptive title. -->

Closes: https://github.com/woocommerce/woocommerce-android/issues/8311
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
There were a few crashes reported (https://github.com/woocommerce/woocommerce-android/issues/8311), potentially happening in case `ShouldShowFeedbackBanner` conditions change during the execution of the target user group detection use case (`GetIPPFeedbackBannerData`). 

While I couldn't reproduce the crash, this PR fixes unsafe error handling design and replaces exception throwing by returning a null value and logging the error, so no crashes are supposed to happen.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
The IPP survey banner should work as expected, and should not cause a crash.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Fix error handling in `GetIPPFeedbackBannerData`

* Don't throw exception in case of illegal state, malformed data or api errors

samiuelson committed:
Update release notes

samiuelson committed:
Merge branch 'trunk' into 8311/ipp-feedback-error-handling

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt:
```diff
@@ -27,8 +27,8 @@ class GetIPPFeedbackBannerData @Inject constructor(
-    suspend operator fun invoke(): IPPFeedbackBanner {
-        requireShouldShowFeedbackBanner()
+    suspend operator fun invoke(): IPPFeedbackBanner? {
+        if (!shouldShowFeedbackBanner()) return null
```
⚠️ We should log this behavior else, we wouldn't know if this is happening for many users out there. Right now, it will fail silently. WDYT?

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/list/OrderListViewModel.kt:
```diff
@@ -183,8 +183,12 @@ class OrderListViewModel @Inject constructor(
                 val bannerData = getIPPFeedbackBannerData()
-                viewState = viewState.copy(ippFeedbackBannerState = IPPSurveyFeedbackBannerState.Visible(bannerData))
-                trackIPPBannerEvent(AnalyticsEvent.IPP_FEEDBACK_BANNER_SHOWN)
+                if (bannerData != null) {
```
❓ What do you think about writing tests to verify this change?

1. Test that ensures that the view state isn't modified when banner data is null
2. Test that ensures that the track event is not called when the banner data is null

AnirudhBhat reviewed (commented):
Thanks for taking care of this @samiuelson. I've left a couple of comments to think about. Let me know your thoughts on those.

samiuelson committed:
Add unit tests for null banner data case

samiuelson committed:
Add missing log

samiuelson committed:
Clean up tests

Removed unnecessary stubbings

samiuelson commented:
> Thanks for taking care of this @samiuelson. I've left a couple of comments to think about. Let me know your thoughts on those.

Thanks, @AnirudhBhat for the review. I've added additional tests to verify correct view states and added the log in case the banner should not be shown.

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/feedback/ipp/GetIPPFeedbackBannerData.kt:
```diff
@@ -28,7 +28,10 @@ class GetIPPFeedbackBannerData @Inject constructor(
     suspend operator fun invoke(): IPPFeedbackBanner? {
-        if (!shouldShowFeedbackBanner()) return null
+        if (!shouldShowFeedbackBanner()) {
+            logger.e(AppLog.T.API, "GetIPPFeedbackBannerData should not be shown.")
```
❓ nitpick:  Sorry for not mentioning this in my first review. I was wondering if it's worth it to write a test to verify that we are logging in case of `shouldShowFeedbackBanner=false`. Same for other loggers as well.

Please feel free to ignore it if you think it's not worth doing this.

AnirudhBhat reviewed (approved):
I've left one minor nitpick that is non-blocking. LGTM :shipit:
